Properly handle a '/' in the cookie's path when setting a cookie with the Cookie Store API

RupinMittal opened the pull request:
#### d52eb029fa40b2be7f7a8275c2e5fccf962a2e7d
<pre>
Properly handle a &apos;/&apos; in the cookie&apos;s path when setting a cookie with the Cookie Store API
<a href="https://bugs.webkit.org/show_bug.cgi?id=259505">https://bugs.webkit.org/show_bug.cgi?id=259505</a>

Reviewed by Chris Dumez and Alex Christensen.

The spec (<a href="https://wicg.github.io/cookie-store/#set-cookie-algorithm)">https://wicg.github.io/cookie-store/#set-cookie-algorithm)</a>
dictates that in the set function, if the path is not null, then if
the path does not begin with a &apos;/&apos;, the promise should be rejected
with a TypeError. If the path does begin with a &apos;/&apos;,  but does not
end with a &apos;/&apos;, then a &apos;/&apos; should be added, and then the rest of the

* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_delete_arguments.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_set_arguments.https.any-expected.txt:
* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::CookieStore::set):

Canonical link: <a href="https://commits.webkit.org/266317@main">https://commits.webkit.org/266317@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/43aad0332d4b5ffc91559b94a2d105ccc7b9f811

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13484 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13798 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/14129 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/15221 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12832 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/13563 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/16306 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13867 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15514 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/13651 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14299 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11416 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/15925 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11586 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/12171 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/19216 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12662 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12339 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15551 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12845 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/10730 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/12114 "Built successfully") | | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3292 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/16441 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12688 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -208,22 +208,22 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
 
+    cookie.path = WTFMove(options.path);
+    if (!cookie.path.isNull()) {
+        if (cookie.path.isEmpty() || cookie.path.characterAt(0) != '/') {
```
cookie.path.startsWith('/')

cdumez replied:
ahah, I was about to suggest the same thing :)

achristensen07 commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -208,22 +208,22 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
+            promise->reject(TypeError);
+            return;
+        }
+        if (cookie.path.characterAt(cookie.path.length() - 1) != '/')
```
cookie.path.endsWith('/')

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -208,22 +208,22 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
+    cookie.path = WTFMove(options.path);
+    if (!cookie.path.isNull()) {
+        if (cookie.path.isEmpty() || cookie.path.characterAt(0) != '/') {
+            promise->reject(TypeError);
```
Please add a description for the exception. How is the web dev supposed to know what there are doing wrong?

cdumez reviewed (approved):
LGTM

achristensen07 commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -208,22 +208,22 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
     cookie.created = WallTime::now().secondsSinceEpoch().milliseconds();
 
+    cookie.path = WTFMove(options.path);
+    if (!cookie.path.isNull()) {
```
I don't think this is needed.

RupinMittal replied:
I may be wrong about this--but if we don't have this check and the path is null, then `cookie.path.startsWith('/')` would be false and we'll return a `TypeError` when we shouldn't right?

achristensen07 replied:
Oh you're right.  This is needed.

RupinMittal committed:
Properly handle a '/' in the cookie's path when setting a cookie with the Cookie Store API
https://bugs.webkit.org/show_bug.cgi?id=259505

Reviewed by Chris Dumez and Alex Christensen.

The spec (https://wicg.github.io/cookie-store/#set-cookie-algorithm)
dictates that in the set function, if the path is not null, then if
the path does not begin with a '/', the promise should be rejected
with a TypeError. If the path does begin with a '/',  but does not
end with a '/', then a '/' should be added, and then the rest of the

* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_delete_arguments.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_set_arguments.https.any-expected.txt:
* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::CookieStore::set):

Canonical link: https://commits.webkit.org/266317@main
